tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -697,23 +697,24 @@ getParamParentNameOffset(const ValueDecl *VD, SourceLoc Cursor) {
   return SM.getLocOffsetInBuffer(Loc, SM.findBufferContainingLoc(Loc));
 }
 
-/// Returns true for failure to resolve.
-static bool passCursorInfoForDecl(SourceFile* SF,
-                                  const ValueDecl *VD,
-                                  const ModuleDecl *MainModule,
-                                  const Type ContainerTy,
-                                  bool IsRef,
-                                  bool RetrieveRefactoring,
-                                  ResolvedCursorInfo TheTok,
-                                  Optional<unsigned> OrigBufferID,
-                                  SourceLoc CursorLoc,
+/// Returns a non-empty StringRef on error, representing an internal diagnostic.
+static StringRef passCursorInfoForDecl(SourceFile* SF,
```

Comment:
It is generally unsafe to return `StringRef` from a free function, it can very easily return a StringRef that points to a released memory location. The pattern we use elsewhere is to have the function accept a `std::string &Error` parameter and when there is a failure return `true` and set the error string, otherwise return `false`.
For this function it could be `std::string &Diagnostic` to make it clear how it is used.

Reply:
Done